### PR TITLE
feat: [CDS-72726]: Disable use of Fabric8 to fetch container info for Native Helm NG

### PR DESCRIPTION
**Pull request by GoelPratyush99**
## Describe your changes
The scope of this PR is limited to Native Helm NG for not relying on fabric8 to perform steady state check and get the container info list. Instead use kubernetes java client to perform steady state for the following workloads- Deployment Config, Deployment, StatefulSets, ReplicaSets, Replication Controller, DaemonSets. (As these were the one supported using fabric8, We do not want to introduce any new feature/ regression with these changes so trying to keep the workflow same for the users.). 

If FF: CDS_DISABLE_FABRIC8_NG is enabled we will not perform steady using fabric8 dependencies but rather use kubernetes java client.

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [x] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/51607)

### CODE REVIEW

**Review by markoharness — COMMENTED** [review]
Please update PR description and testing status

File: 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/apply/K8sApplyBaseHandler.java [hunk 1]
```diff
@@ -120,7 +120,7 @@ public boolean steadyStateCheck(boolean skipSteadyStateCheck, String namespace,
 
       success = k8sTaskHelperBase.doStatusCheckForAllResources(k8sApplyHandlerConfig.getClient(), kubernetesResourceIds,
           k8sDelegateTaskParams, namespace, executionLogCallback,
-          denoteOverallSuccess && k8sApplyHandlerConfig.getCustomWorkloads().isEmpty(), isErrorFrameworkEnabled);
+          denoteOverallSuccess && k8sApplyHandlerConfig.getCustomWorkloads().isEmpty(), isErrorFrameworkEnabled, true);
```
Comment (acasian-harness) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/apply/K8sApplyBaseHandler.java:123]: what is this boolean flag stands for?

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/helm/HelmDeployServiceImplNG.java [hunk 2]
```diff
@@ -473,8 +473,12 @@ private boolean checkIfReleasePurgingNeeded(HelmInstallCommandRequestNG commandR
   public List<ContainerInfo> getContainerInfos(HelmCommandRequestNG commandRequest,
       List<KubernetesResourceId> workloads, boolean useSteadyStateCheck, LogCallback logCallback, long timeoutInMillis)
       throws Exception {
-    return useSteadyStateCheck ? getKubectlContainerInfos(commandRequest, workloads, logCallback, timeoutInMillis)
-                               : getFabric8ContainerInfos(commandRequest, logCallback, timeoutInMillis);
+    if (useSteadyStateCheck) {
+      return getKubectlContainerInfos(commandRequest, workloads, logCallback, timeoutInMillis, true);
+    } else if (commandRequest.isDisableFabric8()) {
+      return getKubectlContainerInfos(commandRequest, workloads, logCallback, timeoutInMillis, false);
```
Comment (acasian-harness) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/task/helm/HelmDeployServiceImplNG.java:479]: i believe the boolean flag stands for `useSteadyStateCheck`? if that's the case we can pass the var itself.

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/helm/HelmDeployServiceImplNG.java [hunk 3]
```diff
@@ -473,8 +473,12 @@ private boolean checkIfReleasePurgingNeeded(HelmInstallCommandRequestNG commandR
...
+    }
+    return getFabric8ContainerInfos(commandRequest, logCallback, timeoutInMillis);
```
Comment (acasian-harness) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/task/helm/HelmDeployServiceImplNG.java:481]: aren't we still fetching container info via fabric8 here: `return getFabric8ContainerInfos(commandRequest, logCallback, timeoutInMillis);` ?

File: 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/apply/K8sApplyBaseHandler.java (hunk 1 above)
Comment (GoelPratyush99) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/apply/K8sApplyBaseHandler.java:123]: It is to status check all the resources. The difference in using the kubernetes java client and fabric 8 implementation in our code base is that for java client we status check for all the resources whereas for fabric8 code flow we only status check deploymentConfig, deployment, replicaSets, DaemonSets, replicationControllers and statefulSets.

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/helm/HelmDeployServiceImplNG.java (hunk 2 above)
Comment (GoelPratyush99) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/task/helm/HelmDeployServiceImplNG.java:479]: Simplified it

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/helm/HelmDeployServiceImplNG.java (hunk 3 above)
Comment (GoelPratyush99) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/task/helm/HelmDeployServiceImplNG.java:481]: Yes we will remove this flow once we GA this FF: **CDS_DISABLE_FABRIC8_NG**

**Comment by GoelPratyush99** [comment]
trigger uts

**Comment by GoelPratyush99** [comment]
trigger uts

**Comment by GoelPratyush99** [comment]
trigger uts

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/k8s/k8sbase/K8sTaskHelperBase.java [hunk 4]
```diff
@@ -1520,18 +1541,16 @@ protected void processLine(String line) {
...
-              executionLogCallback, isErrorFrameworkEnabled);
-        } else {
-          success = doStatusCheckForWorkloads(client, kubernetesResourceId, k8sDelegateTaskParams, statusFormat,
-              executionLogCallback, isErrorFrameworkEnabled);
-        }
-
-        if (!success) {
-          break;
-        }
+      if (!statusCheckAllResources) {
```
Comment (abhinavsingh1196) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/task/k8s/k8sbase/K8sTaskHelperBase.java:1544]: This block is just filtering workloads based on this boolean. This function already accepts the list of workloads to do steady state check for, so we can remove the boolean and do the filtering in the caller ?

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/helm/HelmDeployServiceImplNG.java [hunk 5]
```diff
@@ -543,17 +546,17 @@ && doStatusCheckAllResourcesForHelm(client, entry.getValue(), commandRequest.get
...
+      String gcpKeyFilePath, boolean statusCheckAllResourcesForHelm) throws Exception {
     return k8sTaskHelperBase.doStatusCheckForAllResources(client, resourceIds,
         K8sDelegateTaskParams.builder()
             .ocPath(ocPath)
             .workingDirectory(workingDir)
             .kubeconfigPath(kubeconfigPath)
             .gcpKeyFilePath(gcpKeyFilePath)
             .build(),
-        namespace, executionLogCallback, false, true);
+        namespace, executionLogCallback, false, true, statusCheckAllResourcesForHelm);
```
Comment (abhinavsingh1196) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/task/helm/HelmDeployServiceImplNG.java:559]: could we filter out the resourceIds here itself and remove the boolean `statusCheckAllResourcesForHelm`?

Comment (abhinavsingh1196) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/task/helm/HelmDeployServiceImplNG.java:559]: Also, please add unit tests

File: 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/apply/K8sApplyBaseHandler.java (hunk 1 above)
Comment (abhinavsingh1196) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/apply/K8sApplyBaseHandler.java:123]: +1 for removing this boolean

**Review by abhinavsingh1196 — CHANGES_REQUESTED** [review]

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/k8s/k8sbase/K8sTaskHelperBase.java (hunk 4 above)
Comment (GoelPratyush99) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/task/k8s/k8sbase/K8sTaskHelperBase.java]: Done

File: 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/apply/K8sApplyBaseHandler.java (hunk 1 above)
Comment (GoelPratyush99) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/k8s/apply/K8sApplyBaseHandler.java]: Done

**Comment by GoelPratyush99** [comment]
trigger uts

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/helm/HelmDeployServiceImplNG.java [hunk 6]
```diff
@@ -657,6 +671,17 @@ public HelmCommandResponseNG rollback(HelmRollbackCommandRequestNG commandReques
         rollbackWorkloads = helmSteadyStateService.findEligibleWorkloadIds(resources);
       }
 
+      if (!useSteadyStateCheck && commandRequest.isDisableFabric8()) {
+        List<KubernetesResource> resources = helmSteadyStateService.readManifestFromHelmRelease(
```
Comment (markoharness) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/task/helm/HelmDeployServiceImplNG.java:675]: you dont need to create resources variable. You could directly call

rollbackWorkloads = helmSteadyStateService.readManifestFromHelmRelease(
            HelmCommandDataMapperNG.getHelmCmdDataNG(commandRequest)).stream()....

unless helmSteadyStateService.readManifestFromHelmRelease() can return null and then you need to handle that case

**Comment by GoelPratyush99** [comment]
trigger uts

**Comment by GoelPratyush99** [comment]
trigger uts

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/helm/HelmDeployServiceImplNG.java (hunk 6 above)
Comment (GoelPratyush99) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/task/helm/HelmDeployServiceImplNG.java]: DOne

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/helm/HelmDeployServiceImplNG.java (hunk 5 above)
Comment (GoelPratyush99) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/task/helm/HelmDeployServiceImplNG.java]: done

**Comment by GoelPratyush99** [comment]
trigger uts

**Comment by GoelPratyush99** [comment]
trigger uts

**Review by TarunUba1 — APPROVED** [review]

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/helm/HelmDeployServiceImplNG.java [hunk 7]
```diff
@@ -497,7 +513,8 @@ private List<ContainerInfo> getFabric8ContainerInfos(
     return containerInfos;
   }
 
-  private List<ContainerInfo> getKubectlContainerInfos(HelmCommandRequestNG commandRequest,
+  @VisibleForTesting
```
Comment (abhinavsingh1196) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/task/helm/HelmDeployServiceImplNG.java:516]: nit: this annotation can be removed?

**Review by abosii-harness — CHANGES_REQUESTED** [review]

**Pull request closed without merging**